File: go/state.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
)

type SingleValue struct {
	Value interface{} `json:"Value"`
}

type Update struct {
	Value interface{}
	Diff  int64
}

type State struct {
	state     map[string]int64
	timestamp int64
	valid     bool
	history   []Update
}

func NewState(collectHistory bool) *State {
	state := make(map[string]int64)
	history := []Update{}

	return &State{
		state:     state,
		timestamp: 0,
		valid:     true,
		history:   history,
	}
}

func (s *State) getState() []interface{} {
	list := []interface{}{}

	for key, value := range s.state {
		clone := make(map[string]interface{})
		err := json.Unmarshal([]byte(key), &clone)
		if err != nil {
			fmt.Println(err);
			continue
		}
		for i := int64(0); i < value; i++ {
			list = append(list, clone["Value"])
		}
	}

	return list
}

func (s *State) getHistory() []Update {
	return s.history
}

func (s *State) validate(timestamp int64) error {
	if !s.valid {
		return errors.New("Invalid state.")
	} else if timestamp < s.timestamp {
		s.valid = false
		return errors.New("Invalid timestamp.")
	}
	return nil
}

func (s *State) process(update Update) {
	var sv = SingleValue { Value: update.Value }
	value, err := json.Marshal(sv)

	if err != nil {
		return
	}

	count, ok := s.state[string(value)]
	if !ok {
		count = 0
	}

	count += update.Diff

	if count <= 0 {
		delete(s.state, string(value))
	} else {
		s.state[string(value)] = count
	}

	if s.history != nil {
		s.history = append(s.history, update)
	}
}

func (s *State) Update(updates []Update, timestamp int64) error {
	if len(updates) > 0 {
		err := s.validate(timestamp)
		if err != nil {
			return err
		}
		s.timestamp = timestamp
		for _, update := range updates {
			s.process(update)
		}
	}
	return nil
}

File: rust/src/state.rs
#[derive(Debug, Clone)]
pub(crate) struct Update<T>
where
    T: std::hash::Hash + Eq,
{
    pub(crate) value: T,
    pub(crate) diff: i64,
}

pub(crate) struct State<T>
where
    T: std::hash::Hash + Eq,
{
    state: std::collections::HashMap<T, i64>,
    timestamp: i64,
    valid: bool,
    history: Option<Vec<Update<T>>>,
}

impl<T> State<T>
where
    T: std::hash::Hash + Eq + Clone,
{
    pub(crate) fn new(collect_history: bool) -> State<T> {
        State {
            state: std::collections::HashMap::new(),
            timestamp: 0,
            valid: true,
            history: if collect_history {
                Some(Vec::new())
            } else {
                None
            },
        }
    }

    pub fn get_state(&self) -> Vec<T> {
        let mut list: Vec<T> = Vec::new();

        for (key, value) in &self.state {
            for _ in 0..*value {
                list.push(key.clone());
            }
        }

        list
    }

    pub fn get_history(&self) -> Option<&Vec<Update<T>>> {
        self.history.as_ref()
    }

    fn validate(&self, timestamp: i64) -> Result<(), String> {
        if !self.valid {
            Err("Invalid state.".to_string())
        } else if timestamp < self.timestamp {
            eprintln!("Invalid timestamp.");
            // &mut self.valid = false;
            Err(format!(
                "Update with timestamp ({}) is lower than the last timestamp ({}). Invalid state.",
                timestamp, self.timestamp
            ))
        } else {
            Ok(())
        }
    }

    fn process(&mut self, update: Update<T>) {
        let update_clone = update.clone();
        let value = update.value;
        let diff = update.diff;
        let count = self.state.get(&value).map_or(diff, |&v| v + diff);

        if count <= 0 {
            self.state.remove(&value);
        } else {
            self.state.insert(value, count);
        }

        if let Some(history) = &mut self.history {
            history.push(update_clone);
        }
    }

    pub fn update(&mut self, updates: Vec<Update<T>>, timestamp: i64) -> Result<(), String> {
        if !updates.is_empty() {
            self.validate(timestamp)?;
            self.timestamp = timestamp;
            updates
                .iter()
                .for_each(|update| self.process(update.clone()));
        }
        Ok(())
    }
}
